### PR TITLE
ccp utility

### PR DESCRIPTION
**Pull request by fcangialosi**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -319,6 +319,11 @@ where
 /// Implement this trait and [`portus::Flow`](./trait.Flow.html) to define a CCP congestion control algorithm.
 /// Instances of this type implement functionality which applies to a given
 /// algorithm as a whole.
+/// 
+/// It is recommended that you also implement
```
Comment (akshayknarayan) [inline src/lib.rs]: Instead of this language, can we just write:

> Implement this trait, `portus::CongAlgBuilder`, and `portus::Flow` to define a CCP congestion control algorithm. `portus::CongAlg` and `portus::Flow` interact with the algorithm runtime, and `CongAlgBuilder` specifies how the `impl CongAlg` should be built from given command line arguments.

I think this language makes it more likely people will implement this, even if they just want to do things in their own crate.

File: src/lib.rs [hunk 2]
```diff
@@ -349,6 +354,32 @@ pub trait CongAlg<I: Ipc> {
...
+/// Structs implementing [`portus::CongAlg`](./trait.CongAlg.html) must also implement this trait
+/// (and must be annotated with [`portus_export::register_ccp_alg`]()) 
+/// in order to expose their algorithm for use in other programs, such as the CCP command-line utility.
+///
+/// The expected use of this trait in a calling program is as follows:
+/// ```
+/// let args = CongAlgBuilder::register();
+/// let matches = app.get_matches_from(std::env::args_os());
+/// let alg = CongAlgBuilder::with_arg_matches(matches);
+/// ```
```
Comment (akshayknarayan) [inline src/lib.rs]: does this doctest compile?

File: Cargo.toml [hunk 3]
```diff
@@ -17,16 +17,22 @@ fnv = "1"
 libc = "0.2"
 nix = "0.9.0"
 nom = "^4"
-quote = "0.5"
+quote = "0.6"
```
Comment (akshayknarayan) [inline Cargo.toml]: should this also be a separate commit?

File: src/bin/ccp.rs [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+extern crate quote;
+extern crate regex;
+extern crate structopt;
+extern crate syn;
+extern crate toml;
+extern crate walkdir;
+
+use colored::Colorize;
+use walkdir::WalkDir;
+//use clap::{App, Arg};
```
Comment (akshayknarayan) [inline src/bin/ccp.rs]: remove old code?

File: src/bin/ccp.rs [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+libc = "0.2"
+clap = "2.32"
+portus = "^0.4.3"
+"#
+    );
+
+    for Alg {
+        crate_name,
+        crate_path,
+    } in algs
```
Comment (akshayknarayan) [inline src/bin/ccp.rs]: this formatting seems weird to me, but I guess it makes sense, and might as well follow rustfmt

File: src/bin/ccp.rs [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+    let loads = algs.iter().map(
+        |Alg {
+             crate_name,
+             crate_path: _,
+         }| {
+            let name = Ident::new(crate_name, Span::call_site());
+            let name_key = Ident::new(&format!("{}_key", crate_name), Span::call_site());
+            let name_args = Ident::new(&format!("{}_args", crate_name), Span::call_site());
+            quote! {
+                let #name_key = <#name::__ccp_alg_export as CongAlg<portus::ipc::chan::Socket<portus::ipc::Blocking>>>::name();
```
Comment (akshayknarayan) [inline src/bin/ccp.rs]: yay

File: src/bin/ccp.rs [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+                _ => {
+                    // TODO
+                    eprintln!("error: algorithm '{}' not found. available algorithms are: ", alg_to_run);
+                    return 1;
+                }
+            }
+        }
+    };
+
+    // TODO this doesn't seem to do anything..
```
Comment (akshayknarayan) [inline src/bin/ccp.rs]: ?

File: src/bin/ccp.rs [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+
+    // Our working directory. ~/.ccp unless user says otherwise
+    let root = match opt.dir {
+        Some(p) => p,
+        None => PathBuf::from(format!(
+            "{}/.ccp",
+            std::env::var("HOME").expect("Could not find HOME in environment variables")
+        )),
+    };
+    //println!("==> root directory is {}", root.to_string_lossy().blue().bold());
```
Comment (akshayknarayan) [inline src/bin/ccp.rs]: remove?

File: src/bin/ccp.rs [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+            let argv = std::env::args()
+                .skip_while(|a| a != "--")
+                .collect::<Vec<String>>()[1..]
+                .join(" ");
+            println!("==> running {} {}", alg.bold().blue(), argv);
+            let lib_path = Path::new(&root)
+                .join("lib")
+                .join("target")
+                .join("debug")
+                .join("libccp.so");
```
Comment (akshayknarayan) [inline src/bin/ccp.rs]: does this have to be done per-user btw? It seems like instead, `ccp get` should put this shared library in `/usr/local/ccp` or something, and the library itself can maybe be symlinked to some standard place like `/usr/local/lib`?

File: src/bin/ccp.rs (hunk 5 above)
Comment (fcangialosi) [reply src/bin/ccp.rs]: hm agreed, i didn't do this, rustfmt did :P i don't particularly care either way

File: src/bin/ccp.rs (hunk 9 above)
Comment (fcangialosi) [reply src/bin/ccp.rs]: hm yeah that sounds good, i'll change

File: src/lib.rs (hunk 2 above)
Comment (fcangialosi) [reply src/lib.rs]: hm, it does but to be honest i feel like it shouldn't have lol

Comment (fcangialosi) [reply src/lib.rs]: added no-run
